crates/circuit/src/operations.rs

```diff
@@ -145,6 +146,7 @@ pub trait Operation {
     fn num_clbits(&self) -> u32;
     fn num_params(&self) -> u32;
     fn control_flow(&self) -> bool;
+    fn blocks(&self) -> Vec<CircuitData>;
```

Comment:
Since blocks will only return a populated `Vec` when dealing with a `PyInstruction`, could we make this accept a `py` token regardless? It is very likely that the user here will have obtained the GIL whenever they want to play with an instance of `PyInstruction`, it's all within the name after all. 

I saw your previous comments when discussing with Jake about keeping this consistent. However, in this case, this method is only supposed to work with one type of  operation, one that specifically holds a `PyObject` and is tied quite closely with python, and even then it is not guaranteed to be an instance of `ControlFlowOp`. The function itself will also perform callbacks to python and obtaining the GIL manually might prove itself costly here (it would be nice to benchmark it beforehand to see if this holds true).

Reply:
It's the case that right now `blocks` comes from Python, but once we've ported control flow operations to Rust that goes away. I don't see a compelling reason to ask for a Python token at the `Operation` level for `blocks` but not for its other methods, so I'd rather be consistent to avoid implying to code readers that there's some logical difference between `Operation::blocks` and say, `Operation::matrix`.

In the case of performance, having the GIL already means that `with_gil` will return fastest (as opposed to not holding it). And, we don't even try to acquire it unless this is a control flow operation, so the penalty to calling `with_gil` within `Operation::blocks` would scale with respect to the number of control flow operations in the circuit, rather than the number of op nodes.